python-package/lightgbm/basic.py

```diff
@@ -1853,6 +1854,116 @@ def free_network(self):
         self.network = False
         return self
 
+    def trees_to_dataframe(self):
```

Comment:
@pford221 I've just found a bug, please fix the following edge case:
```
import numpy as np
import lightgbm as lgb

X = np.ones((10, 2))
y = np.random.rand(10)
lgb_data = lgb.Dataset(X, y)

bst = lgb.train({}, lgb_data, num_boost_round=3)
bst.trees_to_dataframe()
```

```
KeyError                                  Traceback (most recent call last)
<ipython-input-5-86ad7722a0da> in <module>
----> 1 bst.trees_to_dataframe()

D:\Miniconda3\lib\site-packages\lightgbm\basic.py in trees_to_dataframe(self)
   1961             model_list.extend(tree_dict_to_node_list(tree['tree_structure'],
   1962                                                      tree_index=tree['tree_index'],
-> 1963                                                      feature_names=feature_names))
   1964 
   1965         return DataFrame(model_list, columns=model_list[0].keys())

D:\Miniconda3\lib\site-packages\lightgbm\basic.py in tree_dict_to_node_list(tree, node_depth, tree_index, feature_names, parent_node)
   1936                                       tree_index=tree_index,
   1937                                       feature_names=feature_names,
-> 1938                                       parent_node=parent_node)
   1939 
   1940             res = [node]

D:\Miniconda3\lib\site-packages\lightgbm\basic.py in create_node_record(tree, node_depth, tree_index, feature_names, parent_node)
   1896             node['tree_index'] = tree_index
   1897             node['node_depth'] = node_depth
-> 1898             node['node_index'] = _get_node_index(tree, tree_index)
   1899             node['left_child'] = None
   1900             node['right_child'] = None

D:\Miniconda3\lib\site-packages\lightgbm\basic.py in _get_node_index(tree, tree_index)
   1879                 is_split = _is_split_node(tree)
   1880                 node_type = 'S' if is_split else 'L'
-> 1881                 node_num = str(tree['split_index' if is_split else 'leaf_index'])
   1882                 return tree_num + node_type + node_num
   1883 

KeyError: 'leaf_index'
```

Reply:
I'm sorry, this comment had to go to the previous my review but GitHub lagged and it hadn't been appended there.

Reply:
OK. This has been handled through a `try...expect` clause. Also, since a single node tree doesn't have a `count` or `weight` keys, I have to deal with this as well. I wrote a robust function to check if each leaf node comes from a single node tree.